File: eventscout/models/event.py
"""
Event domain model for EventScout.
Provides a standardized representation of events collected across all platforms.
"""

from dataclasses import dataclass, field, asdict
from datetime import datetime
from typing import Any, Dict, List, Optional


@dataclass
class Event:
    """
    Standardized Event data model.
    Maintains 100% backward compatibility with existing data/events.json fields
    while providing enriched metadata for discovery, filtering, and notifications.
    """
    # Core identifying fields
    title: str
    event_url: str
    date_time: datetime
    organizer: str
    source: str = "meetup"

    # Location & Attendance
    mode: str = "online"
    mode_location: str = "Online"
    location: Optional[str] = None
    city: Optional[str] = None
    country: Optional[str] = None

    # Media & Display
    poster_image_url: Optional[str] = None
    description: Optional[str] = None

    # Pricing
    is_free: bool = True
    price_amount: Optional[float] = None
    price_currency: Optional[str] = None

    # Platform & Identification
    source_event_id: Optional[str] = None
    registration_url: Optional[str] = None

    # Technical Classification & Discovery
    is_technical: bool = True
    categories: List[str] = field(default_factory=list)

    # Date, Deadlines & Timeline
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    registration_deadline: Optional[datetime] = None
    submission_deadline: Optional[datetime] = None

    # Metadata
    scraped_at: str = field(default_factory=lambda: datetime.utcnow().isoformat() + "Z")

    def __getitem__(self, key: str) -> Any:
        """Enables dictionary-style indexing e.g. event['title']."""
        try:
            return getattr(self, key)
        except AttributeError:
            raise KeyError(key)

    def get(self, key: str, default: Any = None) -> Any:
        """Enables dict-like get method e.g. event.get('price_amount')."""
        return getattr(self, key, default)

    def __contains__(self, key: str) -> bool:
        """Enables 'title' in event checks."""
        return hasattr(self, key)

    def to_dict(self) -> Dict[str, Any]:
        """Convert the Event instance into a JSON-serializable dictionary."""
        d = asdict(self)
        for dt_field in ("date_time", "start_date", "end_date", "registration_deadline", "submission_deadline"):
            val = getattr(self, dt_field, None)
            if isinstance(val, datetime):
                d[dt_field] = val.isoformat()
        return d

File: frontend/src/types/event.ts
export interface Event {
  id?: string;
  _id?: string;
  title: string;
  event_url: string;
  date_time: string;
  organizer: string;
  source: string;
  mode?: string;
  mode_location: string;
  location?: string | null;
  city: string | null;
  country: string | null;
  poster_image_url: string | null;
  description: string | null;
  is_free: boolean;
  price_amount: number | null;
  price_currency: string | null;
  source_event_id: string | null;
  registration_url: string | null;
  is_technical: boolean;
  categories: string[];
  scraped_at: string;

  // Timeline & Deadlines
  start_date?: string | null;
  end_date?: string | null;
  registration_deadline?: string | null;
  submission_deadline?: string | null;
  
  // Intelligent Ranking & Host Reputation Fields
  ranking_score?: number;
  why_recommended?: string[];
  host_tier?: string;
  host_badge?: string;
  host_verified?: boolean;
  score_breakdown?: {
    organization_score?: number;
    quality_score?: number;
    personalization_score?: number;
    urgency_score?: number;
  };
}
